### PR TITLE
[Android] General improvements on tests

### PR DESCRIPTION
**Pull request by almedina-ms**
# Description

* Adds retries to failing interactions with filtered choice sets
* Adds test screenshots as published result of build
* Adds timeout of 1 minute to each test
* Adds a class to enable screenshotting at any point of the test execution

## Stuff not fixed
* The emulator may not start correctly
* The tests may hang at start up

# How Verified

Pipeline was run
It will be very sad if execution fails on CI

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/7162)

### CODE REVIEW

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestWatchRule.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    protected override fun starting(description: Description)
+    {
+        print("Starting test run");
+    }
+
+    protected override fun succeeded(description: Description?) {
+        super.succeeded(description)
+
+        ScreenshotUtil().instance().takeScreenshot(description, true)
```
Comment (licanhua) [inline source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestWatchRule.kt:21]: refactor the component, and a better interface is
```
ScreenshotUtil::takeScreenshot(description, true)
```
or 
```
var instance = new ScreenshotHelper(desscription, true/*fail*/);
instance.takeScreenshot();
```

**Review by jwoo-msft — APPROVED** [review]
looks great with some lines that pop out to me.

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotUtil.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        if (mInstance == null)
+        {
+            mInstance = ScreenshotUtil();
+        }
+
+        return mInstance as ScreenshotUtil;
+    }
+
+    public fun takeScreenshot(screenshotName: String)
```
Comment (licanhua) [inline source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotUtil.kt:24]: A better name would be takeScreenShotAndSaveAs()

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/SaveScreenCaptureProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package io.adaptivecards.uitestapp
+
+import androidx.test.runner.screenshot.ScreenCapture
+import androidx.test.runner.screenshot.ScreenCaptureProcessor
+
+import android.os.Environment
+import java.io.*
+
+
+class SaveScreenCaptureProcessor: ScreenCaptureProcessor
```
[the remark this answers is not in the document]
Comment (almedina-ms) [reply source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/SaveScreenCaptureProcessor.kt:10]: it's kotlin, the expectation is that at some point we migrate all android code to kotlin but we haven't had the time to do it

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt [hunk 4]
```diff
@@ -13,6 +15,8 @@ import io.adaptivecards.uitestapp.ui.inputs.RetrievedInput
 import org.hamcrest.Matchers
 
 object TestHelpers {
+    val TAG = "UITESTING";
```
[the remark this answers is not in the document]
Comment (jwoo-msft) [reply source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt:18]: UI testing

[the remark this answers is not in the document]
Comment (licanhua) [reply source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt:18]: can it be UITesting UI-TESTING

[the remark this answers is not in the document]
Comment (almedina-ms) [reply source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt:18]: UiTesting is right for this one

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotUtil.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import androidx.test.runner.screenshot.ScreenCaptureProcessor
+import androidx.test.runner.screenshot.Screenshot
+import org.junit.runner.Description
+import java.io.IOException
+
+class ScreenshotUtil {
+
+    private var mInstance : ScreenshotUtil? = null;
+
+    public fun instance(): ScreenshotUtil
```
[the remark this answers is not in the document]
Comment (almedina-ms) [reply source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotUtil.kt]: kotlin doesn't have static methods :(

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt [hunk 6]
```diff
@@ -13,39 +15,103 @@ import io.adaptivecards.uitestapp.ui.inputs.RetrievedInput
...
+        val maxRetries = 5
+        var retries = 0
+
+        while (retries < maxRetries) {
+            ++retries
+
+            Log.i(TAG, "$methodName - Try #$retries")
+
+            try {
+                setTextInInput(findInputInValidatedContainer(inputName), hint)
```
Comment (licanhua) [inline source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt:91]: log before action. or update the log info. For example: "Set value to $hint complete"
```                Log.i(TAG, "$methodName - Set value to $hint")
                setTextInInput(findInputInValidatedContainer(inputName), hint)

                Log.i(TAG, "$methodName - Set option to $choiceSetOption")
                selectPopupOption(choiceSetOption)
```

File: source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotHelper.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    // Function to be called from the succeeded or failed methods from a TestWatcher implementation
+    public fun takeScreenshot(description: Description?, success: Boolean) {
+        var suffix = "_success"
+        if (!success) {
+            suffix = "_fail";
+        }
+
+        val testName: String = description!!.methodName + suffix
```
Comment (licanhua) [inline source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/ScreenshotHelper.kt:32]: Don't know which build and which line, I saw 
```
2022-03-24T01:17:34.8578140Z w: /Users/runner/work/1/s/source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/SaveScreenCaptureProcessor.kt: (38, 42): Unnecessary non-null assertion (!!) on a non-null receiver of type ScreenCapture
```

**Review by licanhua — APPROVED** [review]

**Pull request merged**